File: snapOldData.pl
#!/usr/bin/perl
use POSIX;
use File::Basename;

sub mysystem {
    my $cmd = $_[0];
    system($cmd);
    if ($? == -1) {
        return -1;
    }
    elsif ($? & 127) {
        return -1;
    }
    else {
        return ($? >> 8);
    }
}

die "Usage: $0 directory\n"
	unless $#ARGV == 0;
die "\"$ARGV[0]\" is not a directory!\nUsage: $0 directory\n"
	unless -d "$ARGV[0]";
my $workDir = POSIX::tmpnam();
mkdir "$workDir" or die "Couldn't create $workDir ...";
my $dataDir = $ARGV[0];
$dataDir =~ s,/$,,;
my @dirs = split /\//, $dataDir;
my $finalName = $dirs[$#dirs].".signatures.tar.bz2";

my $currentDir = POSIX::getcwd();
chdir $dataDir;
open DIRS, "find . -type d |";
while(<DIRS>) {
    chomp;
    next if /^\.$/ or /^\.\.$/;
    mkdir "$workDir/$_" or die "Failed to create \"$workDir/$_\"\n";
}
close DIRS;
open FILES, "find . -type f |";
while(<FILES>) {
    chomp;
    mysystem("rdiff signature \"$_\" \"$workDir/$_.signature\"");
}
close FILES;
chdir "$workDir";
my $compressedSignatures = POSIX::tmpnam().".tar.bz2";
mysystem("tar jcf \"$compressedSignatures\" ./");
chdir $currentDir;
mysystem("rm -rf \"$workDir\"");
mysystem("mv \"$compressedSignatures\" \"$currentDir/$finalName\"");

File: snapOldData.py
#!/usr/bin/env python
import os
import sys
import shutil
import tarfile
import tempfile

def main():
    if len(sys.argv) != 2:
        sys.stderr.write("Usage: %s directory\n" % sys.argv[0])
        sys.exit(1)
    if not os.path.isdir(sys.argv[1]):
        sys.stderr.write("\"%s\" is not a directory!\nUsage: %s directory\n" \
            (sys.argv[1], sys.argv[0]))
    workDir = tempfile.mkdtemp()
    dataDir = sys.argv[1]
    dataDir = os.path.abspath(sys.argv[1])
    finalName = os.path.abspath(dataDir).split(os.sep)[-1] + ".signatures.tar.bz2"
    tar = tarfile.open(finalName, "w:bz2")

    for direntry in os.walk(dataDir):
        dirpath, dirnames, filenames = direntry
        common = os.path.commonprefix([dirpath, dataDir])
        srcPrefix  = dirpath + os.sep
        destPrefix = workDir + os.sep + dirpath[len(common):] + os.sep
        for dirname in dirnames:
            os.mkdir(destPrefix + dirname)
        for filename in filenames:
            if not os.path.islink(srcPrefix + filename):
                if 0 != os.system("rdiff signature \"%s\" \"%s\"" % \
                        (srcPrefix + filename, destPrefix + filename + ".signature")):
                    sys.exit(1)
                tar.add(\
                    destPrefix + filename + ".signature", \
                    "." + dirpath[len(common):] + os.sep + filename + ".signature")
                print "." + dirpath[len(common):] + os.sep + filename
    tar.close()
    #if 0 != os.system("tar jcf \"%s\" -C \"%s\" ." % (finalName, workDir)):
    #   sys.exit(1)
    shutil.rmtree(workDir, True) # ignore errors in removing

if __name__ == "__main__":
    if sys.argv.count("-pdb") != 0:
        sys.argv.remove("-pdb")
        import pdb
        pdb.run('main()')
    else:
        main()
